contracts/src/SafeInternationalHarbour.sol

```diff
@@ -0,0 +1,406 @@
+     * @param safeAddress Address of the target Safe Smart Account.
+     * @param chainId Chain ID included in the domain separator.
+     * @param nonce Safe transaction nonce.
+     * @param to Target address the Safe will call.
+     * @param value ETH value to be sent with the call.
+     * @param data Call data executed by the Safe.
+     * @param operation Operation type: 0 = CALL, 1 = DELEGATECALL.
+     * @param safeTxGas Gas limit for the Safe's internal execution.
+     * @param baseGas Base gas overhead for reimbursement.
+     * @param gasPrice Gas price used for reimbursement calculation.
+     * @param gasToken Token address for refunds (0x0 for ETH).
+     * @param refundReceiver Address to receive gas refunds.
+     * @return safeTxHash Keccak256 digest of the EIP-712 encoded SafeTx.
+     */
+    function _computeSafeTxHash(
+        address safeAddress,
+        uint256 chainId,
+        uint256 nonce,
+        address to,
+        uint256 value,
+        bytes calldata data,
+        uint8 operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address refundReceiver
+    ) private pure returns (bytes32 safeTxHash) {
+        bytes32 domainSeparator = keccak256(
+            abi.encode(_DOMAIN_TYPEHASH, chainId, safeAddress)
+        );
+        bytes32 structHash = keccak256(
+            abi.encode(
+                _SAFE_TX_TYPEHASH,
+                to,
+                value,
+                keccak256(data),
+                operation,
+                safeTxGas,
+                baseGas,
+                gasPrice,
+                gasToken,
+                refundReceiver,
+                nonce
+            )
+        );
+        safeTxHash = keccak256(abi.encodePacked("\x19\x01", domainSeparator, structHash));
+    }
+
+    /**
+     * @notice Splits a 65-byte ECDSA signature into its components and recovers the signer address.
+     * @param digest The message or data hash to verify (EIP-712 digest or eth_sign prefixed).
+     * @param sig Concatenated 65-byte ECDSA signature (r || s || v).
+     * @return signer The address that produced the signature (EOA).
+     * @return r First 32 bytes of the ECDSA signature.
+     * @return s Second 32 bytes of the ECDSA signature.
+     * @dev Supports both EIP-712 and eth_sign flows by detecting v > 30 and applying the Ethereum Signed Message prefix.
+     */
+    function _recoverSignerAndRS(bytes32 digest, bytes calldata sig)
+        private
+        pure
+        returns (address signer, bytes32 r, bytes32 s)
+    {
+        uint8 v;
+        assembly ("memory-safe") {
+            r := calldataload(sig.offset)
+            s := calldataload(add(sig.offset, 0x20))
+            v := byte(0, calldataload(add(sig.offset, 0x40)))
+        }
+        if (v > 30) {
+            digest = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", digest));
+            v -= 4;
```

Comment:
The issue with supporting this, is that you need to "guess" what format was used when recovering. Not sure if it is an issue.

My intuition is that it isn't. If you pack the `v` with the `s`, then you only need to test 2 `ecrecover`s per signature, and can even do it on-chain in a view function.

Reply:
Another alternative is to _just_ support EIP-712 signatures. IMO we should push for this anyway (as we don't want to support blind-signing).

Reply:
> Another alternative is to just support EIP-712 signatures. IMO we should push for this anyway (as we don't want to support blind-signing).

I also thought about this. I'd say this is the way. Will adjust